File: src/processor.rs
use std::time::SystemTime;

use opentelemetry::global::ObjectSafeSpan;
use opentelemetry_sdk::trace::SpanProcessor;

use crate::wit::v2::observe;

#[derive(Debug)]
pub struct WasiProcessor {}

impl WasiProcessor {
    pub fn new() -> Self {
        Self {}
    }
}

impl Default for WasiProcessor {
    fn default() -> Self {
        Self::new()
    }
}

impl SpanProcessor for WasiProcessor {
    fn on_start(&self, span: &mut opentelemetry_sdk::trace::Span, _cx: &opentelemetry::Context) {
        println!(
            "Processor on_start passing span_id: {:?}",
            span.span_context().span_id().to_string()
        );
        observe::on_span_start(&observe::SpanContext {
            span_id: span.span_context().span_id().to_string(),
            trace_id: span.span_context().trace_id().to_string(),
            trace_flags: format!("{:x}", span.span_context().trace_flags()),
            is_remote: span.span_context().is_remote(),
            trace_state: "".to_string(), // TODO
        })
    }

    fn on_end(&self, span: opentelemetry_sdk::export::trace::SpanData) {
        println!("PROCESSOR ON_END FOR SPAN: {:?}\n\n", span.name.to_string());
        println!(
            "Processor on_end passing span_id: {:?}",
            span.span_context.span_id().to_string()
        );
        let start_since_the_epoch = span
            .start_time
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap();
        let end_since_the_epoch = span
            .end_time
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap();
        observe::on_span_end(&observe::ReadOnlySpan {
            name: span.name.to_string(),
            span_context: observe::SpanContext {
                span_id: span.span_context.span_id().to_string(),
                trace_id: span.span_context.trace_id().to_string(),
                trace_flags: format!("{:x}", span.span_context.trace_flags()),
                is_remote: span.span_context.is_remote(),
                trace_state: "".to_string(), // TODO
            },
            parent_span_id: span.parent_span_id.to_string(),
            span_kind: observe::SpanKind::Internal, // TODO
            start_time: observe::Datetime {
                seconds: start_since_the_epoch.as_secs(),
                nanoseconds: start_since_the_epoch.subsec_nanos(),
            },
            end_time: observe::Datetime {
                seconds: end_since_the_epoch.as_secs(),
                nanoseconds: end_since_the_epoch.subsec_nanos(),
            },
            attributes: vec![],
            otel_resource: observe::OtelResource {
                attrs: vec![],
                schema_url: None,
            },
        })
    }

    fn force_flush(&self) -> opentelemetry::trace::TraceResult<()> {
        Ok(())
    }

    fn shutdown(&mut self) -> opentelemetry::trace::TraceResult<()> {
        Ok(())
    }
}

File: go/processor.go
package wasi_otel

import (
	"context"
	"fmt"
	"sync/atomic"

	wasiTrace "github.com/calebschoepp/opentelemetry-wasi/internal/wasi/otel/tracing"
	sdkTrace "go.opentelemetry.io/otel/sdk/trace"
)

// Source: https://github.com/open-telemetry/opentelemetry-go/blob/20cd7f871fce8fae83ea35958775043d6e94b925/sdk/trace/span_processor.go#L15

type WasiProcessor struct {
	IsShutdown atomic.Bool
}

func NewWasiProcessor() *WasiProcessor {
	return &WasiProcessor{}
}

func DefaultWasiProcessor() *WasiProcessor {
	return NewWasiProcessor()
}

func (p *WasiProcessor) OnStart(parent context.Context, s sdkTrace.ReadWriteSpan) {
	// TODO: `parent` seems never to be used...
	if p.IsShutdown.Load() {
		return
	}

	wasiTrace.OnStart(toWasiSpanContext(s.SpanContext()))
}

func (p *WasiProcessor) OnEnd(s sdkTrace.ReadOnlySpan) {
	if p.IsShutdown.Load() {
		return
	}

	wasiTrace.OnEnd(wasiTrace.SpanData{
		SpanContext:          toWasiSpanContext(s.SpanContext()),
		ParentSpanID:         s.Parent().SpanID().String(),
		SpanKind:             wasiTrace.SpanKind(s.SpanKind()),
		Name:                 s.Name(),
		StartTime:            toWasiDateTime(s.StartTime()),
		EndTime:              toWasiDateTime(s.EndTime()),
		Attributes:           toWasiAttributes(s.Attributes()),
		Events:               toWasiEvents(s.Events()),
		Links:                toWasiLinks(s.Links()),
		Status:               toWasiStatus(s.Status()),
		InstrumentationScope: toWasiInstrumentationScope(s.InstrumentationScope()),
		DroppedAttributes:    uint32(s.DroppedAttributes()),
		DroppedEvents:        uint32(s.DroppedEvents()),
		DroppedLinks:         uint32(s.DroppedLinks()),
	})
}

func (p *WasiProcessor) ForceFlush(ctx context.Context) error {
	if p.IsShutdown.Load() {
		return fmt.Errorf("processor already shutdown")
	}

	return nil
}

func (p *WasiProcessor) Shutdown(ctx context.Context) error {
	err := p.ForceFlush(ctx)
	if p.IsShutdown.Swap(true) {
		return err
	}

	return nil
}
